src/tools/clippy/clippy_lints/src/missing_const_for_fn.rs

```diff
@@ -108,36 +107,21 @@ impl<'tcx> LateLintPass<'tcx> for MissingConstForFn {
             FnKind::Method(_, sig, ..) => {
                 if trait_ref_of_method(cx, hir_id).is_some()
                     || already_const(sig.header)
-                    || method_accepts_dropable(cx, sig.decl.inputs)
                 {
                     return;
                 }
             },
             FnKind::Closure(..) => return,
         }
 
-        let mir = cx.tcx.optimized_mir(def_id);
+        let mir = cx.tcx.mir_const(WithOptConstParam::unknown(def_id)).borrow();
```

Comment:
whoa, how does this work? Couldn't it already have been stolen?

Reply:
Clippy tests don't run on PR CI, and I still can't run them locally due to some error about "two copies of the `regex` crate".

I don't know if clippy has a mechanism for ordering lints. It's not correct to const-check optimized MIR, since optimizations (especially drop elaboration) could have removed invalid operations. This was true for `qualify_min_const_fn` as well. Can someone from @rust-lang/clippy weigh in here?

Reply:
Okay, `./x.py clean` seemed to work. Besides being incorrect, running const-checking on optimized MIR causes a bunch of assertions to fire, since we don't reason about promoteds inside const-checking. I guess we'll have to disable the lint until someone tells us how to order queries across clippy lints.

Reply:
clippy could override the query that steals `mir_const` and just not steal it. Right now we're doing a hacky thing (https://github.com/rust-lang/rust/blob/87d262acb50200d767baa7115f30c650a13672ee/src/tools/clippy/src/driver.rs#L86) but maybe we could just inject a different query.

cc @rust-lang/clippy are you ok with turning off the "this could be const fn" lint until we figure out how to get it to use exactly the same analysis as the one used by rustc?

Reply:
This passes clippy tests locally now that I've commented out a single assertion. It was just the one ICE in every test. I think we could live with it (that assertion was more of a sanity check), but it doesn't feel great.

Reply:
@oli-obk seems okay to me!

Reply:
> are you ok with turning off the "this could be const fn" lint

totally - I would still consider it experimental anyway. It's in the nursery for good reason :)
I guess we could temporarily deprecate it, if needed?

Reply:
Yea I don't like adding hacks to those checks just so clippy can (ab)use that API. As I mentioned in https://github.com/rust-lang/rust/pull/77170#issuecomment-698851063 we could just move the old checker to clippy. This means clippy won't automatically catch up with rustc, but at least we're keeping the status quo of suggesting simple functions to be turned into const fn

Reply:
> I guess we could temporarily deprecate it, if needed?

I don't think we should do this. Since it is already in nursery, we could just empty out the lint pass of the lint and just revert this commit once we figured it out.

@ecstatic-morse Clippy is tested by running `./x.py test src/tools/clippy`

---

Also Clippy has set the `mir-opt`-flag to `0`, so no MIR optimizations are run for Clippy lints.

Reply:
> Clippy is tested by running ./x.py test src/tools/clippy

If it works like Miri, `test --stage 0` should also work (and be much faster as rustc is built only once).

Reply:
I tried started with `stage 0` but got the bug about two rlibs for `regex`. Can anyone reproduce? Doing a stage 1 build just to run clippy tests is pretty painful, especially since they don't run on CI.

Reply:
the regex thing is a longstanding problem with compiletest-rs that we've not really been able to fix. I don't know if https://crates.io/crates/trybuild fares better, maybe we should try it out

Reply:
Miri also uses compiletest and its stage 0 tests work fine...

Reply:
in miri we don't use any crates in our unit tests I think